src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
-    IsLast _ = 'False
+{- | Constraint for `PWrapLam'` that will give a better type error message.
+
+ @since 2.1.0
+-}
+type PWrapLam (h :: Type) = (PWrapLam' h (IsLam h) (IsLast h), OnlyTestableTerm h)
 
-{- | Finds Haskell level TestableTerm equivlance of Plutarch
-     functions. This TypeFamily expects the input Plutarch functions to
-     be returning @PBool@ at the end.
+{- | Replace any TestableTerm that is a lambda with `PFun` for generation and
+     shrinking. It will ignore lambda on the return as it should not be
+     generated.
 
-     This is used to find type signatures for @quickCheck@able
-     functions from Plutarch terms like @Term s (a :--> b :--> PBool)@.
+= Note
+"TestableTerm that is lambda" means @TestableTerm (a :--> b)@.
 
- @since 2.0.0
+ @since 2.1.0
 -}
-type family TestableFun (b :: PTermType) (p :: S -> Type) where
-    TestableFun _ PBool = TestableTerm PBool
-    TestableFun 'PTerm (a :--> b) = TestableTerm a -> TestableFun (PTT b) b
-    TestableFun 'LastPTerm (a :--> b) = TestableTerm a -> TestableFun (PTT b) b
-    TestableFun 'PFunction ((a :--> b) :--> c) =
-        PFun a b -> TestableFun (PTT c) c
-    TestableFun 'LastPFunction ((a :--> b) :--> c) =
-        PFun a b -> TestableFun (PTT c) c
-
-{- | Convert Plutarch function into testable Haskell function that takes
-     `TestableTerm`. It also converts plutarch function into `PFun`.
-
- @since 2.0.0
+pwrapLam ::
+    forall (h :: Type).
+    PWrapLam h =>
+    h ->
+    PLamWrapped h
+pwrapLam = pwrapLam' @h @(IsLam h) @(IsLast h)
+
+-- Get the type of Haskell Testable function from given Plutarch Function.
+-- It requires the final return type and the actual function type, both in PType.
+--
+-- This typeclass is not exhaustive: it will not reduce when provided final return
+-- type is not an actual return type of Plutarch function.
+type family PUnLamHask (fin :: S -> Type) (p :: S -> Type) :: Type where
+    PUnLamHask a a = TestableTerm a
+    PUnLamHask fin ((a :--> b) :--> c) =
+        TestableTerm (a :--> b) -> PUnLamHask fin c
+    PUnLamHask fin (a :--> b) = TestableTerm a -> PUnLamHask fin b
+
+-- Helper for type error.
+-- This will give a type error when final return type given does not match
+-- that of the actual function.
+type family CheckReturn (fin :: S -> Type) (p :: S -> Type) :: Constraint where
+    CheckReturn a a = ()
+    CheckReturn fin ((a :--> b) :--> c) = CheckReturn fin c
+    CheckReturn fin (a :--> b) = CheckReturn fin b
+    CheckReturn a b =
+        ( a ~ b
+        , TypeError
+            ( 'Text "Return type does not match:"
+                ':$$: 'Text "\tExpected \""
+                ':<>: 'ShowType a
+                ':<>: 'Text "\" but given function returns \""
+                ':<>: 'ShowType b
+                ':<>: 'Text "\""
+            )
+        )
+
+type family IsFinal (fin :: S -> Type) (p :: S -> Type) :: Bool where
+    IsFinal a a = 'True
+    IsFinal _ _ = 'False
```

Comment:
Isn't this just [`:~:`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Type-Equality.html#t::-126-:)?